### DISCUSSION TITLE
Usage of `minify`

### DISCUSSION THREAD

**Discussion by Malix-Labs** [lisonge/vite-plugin-monkey#55] [Q&A] [answered]
Do you know if there is a way to make the userscript work if `build.minify = true`?

**Comment by lisonge** [comment]
if `build.minify = true`, Why don't the build work  ?

**Reply by Malix-Labs** [reply]
The build works, but is unusable by Tampermonkey

**Reply by lisonge** [reply]
what is the error console log in Tampermonkey

and can you show the code after the build ?

you can show it by <https://gist.github.com/>

**Reply by Malix-Labs** [reply]
- [Comment Reply](https://github.com/lisonge/vite-plugin-monkey/discussions/55#discussioncomment-5273336) from @lisonge:
  > what is the error console log in Tampermonkey [⋯](https://github.com/lisonge/vite-plugin-monkey/discussions/55#discussioncomment-5273336)

  [![tampermonkey error lint](https://user-images.githubusercontent.com/76160668/224485544-90e41ef3-8a48-47d6-a317-c3719d440b1e.png "tampermonkey error lint")*tampermonkey error lint*](https://gist.github.com/Malix-Off/980d720e0d38cf17f5581f4537e163e2/raw/90dcdcc05f07bde8997d7501fa7a3e108a6c66f9/build_minified.user.js)

  ![console error log](https://user-images.githubusercontent.com/76160668/224486552-665f82ba-f6c0-4fec-af8d-4ea82495dba2.png "console error log")*console error log*

  ![console error log﹣file](https://user-images.githubusercontent.com/76160668/224486593-77b3882e-e098-4552-a1bd-18a39b562362.png "console error log﹣file")*console error log﹣file*

- [Comment Reply](https://github.com/lisonge/vite-plugin-monkey/discussions/55#discussioncomment-5273336) from @lisonge:
  > [⋯](https://github.com/lisonge/vite-plugin-monkey/discussions/55#discussioncomment-5273336) can you show the code after the build ? [⋯](https://github.com/lisonge/vite-plugin-monkey/discussions/55#discussioncomment-5273336)

  **[minified build file output](https://gist.github.com/Malix-Off/980d720e0d38cf17f5581f4537e163e2#file-build_minified-user-js)**

## Further

### Links

- **[normal build file output](https://gist.github.com/Malix-Off/980d720e0d38cf17f5581f4537e163e2#file-build_normal-user-js)**
- **[build file source](https://gist.github.com/Malix-Off/980d720e0d38cf17f5581f4537e163e2#file-source-ts)**

**Reply by lisonge** [reply]
**it may be a bug of tampermonkey**

error example

```js
// ==UserScript==
// @name         New Userscript
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @match        https://*
// @grant        GM_getValue
// ==/UserScript==

(function () {
    'use strict';
    var Q = window,
        mt = (() => Q.GM_getValue)();
    console.log([Q,mt]);
})();
```
![image](https://user-images.githubusercontent.com/38517192/224490268-6572e22a-65d3-45af-ba03-95d930f2dea5.png)

---
work axample，`just add a comment`
```js
// ==UserScript==
// @name         New Userscript
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @match        https://*
// @grant        GM_getValue
// ==/UserScript==

(function () {
    'use strict';
    var Q = window,
        mt = (() => Q.GM_getValue)();
    console.log([Q,mt]);
    // window.GM_getValue
})();

```
![image](https://user-images.githubusercontent.com/38517192/224490328-34effeb7-e06c-4231-b52a-9506131b1b80.png)

I once mentioned this bug to the author, [tampermonkey#1567](https://github.com/Tampermonkey/tampermonkey/issues/1567)

I didn't want to care about it at the time, so I closed it directly

**Reply by lisonge** [reply]
you can use [violentmonkey](https://github.com/violentmonkey/violentmonkey)

It's also a userscript engine like tampermonkey, and it's open source

--- 

or use [Global variables usage](https://github.com/lisonge/vite-plugin-monkey#global-variables-usage)

change vite-ent.d.ts file content  to the following code for global GM_api type hint

```ts
/// <reference types="vite/client" />
//// <reference types="vite-plugin-monkey/client" />
/// <reference types="vite-plugin-monkey/global" />
```

**Reply by lisonge** [reply]
Other ways to reduce the size of the output file

```txt
// openai dependencies
externalGlobals: {
    axios: cdn.jsdelivr(`axios`, `dist/axios.min.js`),
    'form-data': 'FormData',
},
```

**Reply by Malix-Labs** [reply]
> Other ways to reduce the size of the output file
> 
> ```
> // openai dependencies
> externalGlobals: {
>     axios: cdn.jsdelivr(`axios`, `dist/axios.min.js`),
>     'form-data': 'FormData',
> },
> ```

Where should I put that?
Would it not actually make it even larger, because vite does tree-shake, but "[`@require`](https://www.tampermonkey.net/documentation.php?locale=en#meta:require)" just straight import the whole source file

**Reply by lisonge** [reply]
>vite does tree-shake

 yes, but no tree-shake for axios.
and there is global object named FormData in Brower for form-data , you dont need import it when you use openai in Brower instead of node

**Reply by Malix-Labs** [reply]
- 
  > no tree-shake for axios

  Do you mean that axios isn't tree-shakable?

-
  > and there is global object named FormData in Brower for form-data , you dont need import it when you use openai in Brower instead of node

  So it will import axios only once, not every time I reload the page?
  If yes, when does it clear the object?

**Reply by Malix-Labs** [reply]
> > Other ways to reduce the size of the output file
> > ```
> > // openai dependencies
> > externalGlobals: {
> >     axios: cdn.jsdelivr(`axios`, `dist/axios.min.js`),
> >     'form-data': 'FormData',
> > },
> > ```
> 
> Where should I put that?

**Reply by lisonge** [reply]
> Where should I put that?

[#cdn-util-for-external](https://github.com/lisonge/vite-plugin-monkey#cdn-util-for-external)

---

>Do you mean that axios isn't tree-shakable?

yes, there is  only one export variable in axios, you can check it in [axios@0.27.2/lib/axios.js](https://unpkg.com/browse/axios@0.27.2/lib/axios.js), [axios@1.3.4/lib/axios.js](https://unpkg.com/browse/axios@1.3.4/lib/axios.js)

---

>So it will import axios only once, not every time I reload the page?
If yes, when does it clear the object?

openai is only a Node.js Library, there is not [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData#browser_compatibility) in nodejs if Nodejs version is less than 18, so openai polyfill it by [form-data](https://www.npmjs.com/package/form-data)  

but there is global  FormData in browser, so we should just external `form-data` to `FormData` without require any url

**Reply by Malix-Labs** [reply]
How would that reduce the total load required ?

**Reply by lisonge** [reply]
>the total load required

what is this ?

**Reply by Malix-Labs** [reply]
> >the total load required
> 
> what is this ?

By "Total Load", I meant the total amount of code fetched/bundled

**Reply by lisonge** [reply]
I can't say clearly

You can see the changes in the building products to understand it

**Comment by Malix-Labs** [comment, marked as answer]
See #56
